mathoxide-lib/src/views.rs

```diff
@@ -1,47 +1,77 @@
 pub trait ArrayView {
     fn translate<ListType: AsRef<[usize]>>(&self, idx: ListType) -> usize;
     fn offset(&self) -> usize;
-    fn ndims(&self) -> usize;
-    fn size(&self) -> usize;
     fn shape(&self) -> &[usize];
+    fn stride(&self) -> &[usize];
+    fn ndim(&self) -> usize {
+        self.shape().len()
+    }
+    fn numel(&self) -> usize {
+        self.shape().iter().product()
+    }
+    fn is_contiguous(&self) -> bool;
 }
 
-pub struct SimpleView {
+pub struct ContiguousView {
     shape: Vec<usize>,
+    offset: usize,
+    stride: Vec<usize>,
 }
 
-impl SimpleView {
+impl ContiguousView {
     pub fn new<ListType: AsRef<[usize]>>(shape: ListType) -> Self {
+        Self::new_with_offset(shape, 0)
+    }
+
+    pub fn new_with_offset<ListType: AsRef<[usize]>>(shape: ListType, offset: usize) -> Self {
         Self {
             shape: shape.as_ref().to_vec(),
+            offset,
+            stride: Self::compute_stride(shape.as_ref()),
         }
     }
+
+    fn compute_stride<ListType: AsRef<[usize]>>(shape: ListType) -> Vec<usize> {
+        let mut res = shape
+            .as_ref()
+            .iter()
+            .rev()
+            .scan(1, |state, &x| {
+                let tmp = *state;
+                *state = *state * x;
```

Comment:
clippy complains for this line and suggets the alternative:

```
*state *= x;
```

even though the official docs don't use this style. I would suggest sticking with clippy on this one.